Userland/Libraries/LibLocate/Crawler.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2021, Aron Lander <aron@aronlander.se>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibLocate/Crawler.h>
+
+namespace Locate {
+
+Crawler::Crawler(String path)
+{
+    m_dir_queue.enqueue(make<Locate::DirectoryInfo>(Locate::DirectoryInfo {
+        .path = path,
+        .db_id = m_id_counter,
+        .parent_db_id = 0,
+        .children = SinglyLinkedListWithCount<NonnullOwnPtr<Locate::ChildInfo>>() }));
+    m_id_counter++;
+}
+
+NonnullOwnPtr<Locate::DirectoryInfo> Crawler::index_next_directory()
+{
+    auto cur_dir_info = m_dir_queue.dequeue();
+    String cur_path = cur_dir_info->path.characters();
+    Core::DirIterator dir_iterator(cur_path, Core::DirIterator::SkipParentAndBaseDir);
+
+    while (dir_iterator.has_next()) {
+        NonnullOwnPtr<Locate::ChildInfo> child = make<Locate::ChildInfo>();
+        child->name = dir_iterator.next_path();
+
+        String fullpath;
+        if (cur_path == "/")
+            fullpath = String(String::formatted("{}{}", cur_path.characters(), child->name.characters()));
+        else
+            fullpath = String(String::formatted("{}/{}", cur_path.characters(), child->name.characters()));
+
+        struct stat stat_info;
+        lstat(fullpath.characters(), &stat_info);
+        child->db_id = m_id_counter;
+        child->parent_db_id = cur_dir_info->db_id;
+        child->user_owner = stat_info.st_uid;
+        child->group_owner = stat_info.st_gid;
+        child->st_mode = stat_info.st_mode & ~0xFE00; //Only keep the permission bits
```

Comment:
```suggestion
        child->st_mode = stat_info.st_mode & ~0xFE00; // Only keep the permission bits
```

Reply:
Well, this is certainly a case of "why did I even implement that?". So, initially, I thought of storing the permissions and checking them upon loading the directory, but later noticed the "access"-function, which made life a lot easier. Anyway, I have some cleaning up to do here (the whole permission bit masking and storing it in the db is totally unnecessary), since this will result in the locatedb storing a lot of unnecessary data.